## Installation note

This notebook demonstrates usage of the `postlink` package.

Because `postlink` depends on `rstan`, installation requires a local R environment with the package installed. In some hosted Jupyter or Colab environments, this step may require additional setup.

In [1]:
# Data input: aggregated counts only
agg <- data.frame(
  NRSHOM     = as.integer(c(0, 1, 0, 1, 0, 0)),
  mds_nrshom = as.integer(c(0, 0, 1, 1, 0, 1)),
  mismatch   = as.integer(c(0, 0, 0, 0, 1, 1)),
  count      = as.integer(c(271, 9, 5, 15, 56, 3))
)

# Display aggregated data
str(agg)
agg

'data.frame':	6 obs. of  4 variables:
 $ NRSHOM    : int  0 1 0 1 0 0
 $ mds_nrshom: int  0 0 1 1 0 1
 $ mismatch  : int  0 0 0 0 1 1
 $ count     : int  271 9 5 15 56 3


NRSHOM,mds_nrshom,mismatch,count
<int>,<int>,<int>,<int>
0,0,0,271
1,0,0,9
0,1,0,5
1,1,0,15
0,0,1,56
0,1,1,3


# 1. Linked Dataset: HRS–Medicare Claims Data

This notebook uses an aggregated HRS-CMS demonstration dataset created directly in the notebook. Each row represents one combination of HRS self-report, CMS administrative record, and linkage mismatch status, together with the number of records in that cell.

The variables in the aggregated dataset are:

| Variable       | Meaning                                        | Values                        |
| -------------- | ---------------------------------------------- | ----------------------------- |
| **NRSHOM**     | HRS self-report of nursing home residence      | 1 = yes, 0 = no               |
| **mds_nrshom** | CMS record of nursing home residence           | 1 = yes, 0 = no               |
| **mismatch**   | Indicator of linkage mismatch status           | 1 = mismatch, 0 = exact match |
| **count**      | Number of records in each aggregated cell      | positive integer              |

The overall linkage mismatch rate is computed as the weighted proportion of records with `mismatch = 1`. Because the data are aggregated, the mismatch indicator is weighted by the cell count.

In [14]:
# Overall linkage mismatch rate using aggregated counts
mismatch_count <- sum(agg$mismatch * agg$count)
total_count <- sum(agg$count)
mm_rate <- mismatch_count / total_count

mismatch_summary <- data.frame(
  Total_records = total_count,
  Mismatched_records = mismatch_count,
  Mismatch_rate = round(mm_rate, 4)
)

mismatch_summary

Total_records,Mismatched_records,Mismatch_rate
<int>,<int>,<dbl>
359,59,0.1643


The estimated mismatch rate is 0.1643, meaning that approximately 16.43% of records in the aggregated demonstration data are treated as linkage mismatches. This motivates the adjusted analysis in Section 2.3.

# 2. Contingency Table Analysis

### 2.1 Naive Approach (compute observed contingency table)

We compute the observed contingency table using all linked records, ignoring potential mismatch error.

In [2]:
# Total number of records represented in the aggregated data
n <- sum(agg$count)

# Naive contingency table in counts
ctab <- xtabs(count ~ NRSHOM + mds_nrshom, data = agg)
ctab

# Naive contingency table in proportions
ctab_prop <- ctab / sum(ctab)
round(ctab_prop, 4)

      mds_nrshom
NRSHOM   0   1
     0 327   8
     1   9  15

      mds_nrshom
NRSHOM      0      1
     0 0.9109 0.0223
     1 0.0251 0.0418

### 2.2 Exact Approach (restrict to correct matches)

To estimate the true distribution, we restrict the dataset to correctly matched records (`mismatch = 0`).

In [3]:
# Keep only correctly linked records
agg_exact <- subset(agg, mismatch == 0)

# Exact-match contingency table in counts
ctab_exact <- xtabs(count ~ NRSHOM + mds_nrshom, data = agg_exact)
ctab_exact

# Exact-match contingency table in proportions
ctab_exact_prop <- ctab_exact / sum(ctab_exact)
round(ctab_exact_prop, 4)

      mds_nrshom
NRSHOM   0   1
     0 271   5
     1   9  15

      mds_nrshom
NRSHOM      0      1
     0 0.9033 0.0167
     1 0.0300 0.0500

### 2.3 Adjustment Using postlink (create adjustment object, compute adjusted table)

We now apply the high-level `postlink` workflow to adjust the observed contingency table for mismatch error using the observed mismatch rate.
Because `postlink` depends on `rstan`, this step requires a local R environment with the package installed and may not run inside hosted Jupyter environments.

In [12]:
suppressPackageStartupMessages(
  suppressWarnings(library(postlink))
)

# Expand aggregated counts only for compatibility with postlink
hrscms_clean <- agg[
  rep(seq_len(nrow(agg)), agg$count),
  c("NRSHOM", "mds_nrshom")
]

rownames(hrscms_clean) <- NULL

# Mismatch rate from aggregated counts
mm_rate <- sum(agg$mismatch * agg$count) / sum(agg$count)

# Create adjustment object
adj <- adjMixture(
  linked.data = hrscms_clean,
  m.rate = mm_rate
)

# Adjusted contingency table
adjusted_table <- plctable(
  formula = ~ NRSHOM + mds_nrshom,
  adjustment = adj
)

# Adjusted probability table
round(adjusted_table$phat, 4)

      mds_nrshom
NRSHOM      0      1
     0 0.9139 0.0170
     1 0.0197 0.0494

In [7]:
# Alternative lower mismatch rate
lower_mm_rate <- 0.5 * mm_rate

adj_lower <- adjMixture(
  linked.data = hrscms_clean,
  m.rate = lower_mm_rate
)

adjusted_table_lower <- plctable(
  formula = ~ NRSHOM + mds_nrshom,
  adjustment = adj_lower
)

# Adjusted probability table using lower mismatch rate
round(adjusted_table_lower$phat, 4)

      mds_nrshom
NRSHOM      0      1
     0 0.9132 0.0194
     1 0.0222 0.0452

# 3. Comparison of Results

Side-by-side comparison of the naive, exact-match, and adjusted contingency tables.  
The adjusted `postlink` workflow is shown in Section 2.3 as the intended package-level analysis. The comparison below summarizes the naive, exact-match, and adjusted contingency tables.

In [8]:
# Side-by-side comparison of naive, exact-match, and adjusted tables
comparison_results <- data.frame(
  Naive   = as.vector(t(ctab_prop)),
  Exact   = as.vector(t(ctab_exact_prop)),
  Adjusted = as.vector(t(adjusted_table$phat)),
  Adj_Lower_Mismatch = as.vector(t(adjusted_table_lower$phat))
)

row.names(comparison_results) <- c(
  "HRS0_CMS0",
  "HRS0_CMS1",
  "HRS1_CMS0",
  "HRS1_CMS1"
)

round(comparison_results, 4)

,Naive,Exact,Adjusted,Adj_Lower_Mismatch
,<dbl>,<dbl>,<dbl>,<dbl>
HRS0_CMS0,0.9109,0.9033,0.9139,0.9132
HRS0_CMS1,0.0223,0.0167,0.0170,0.0194
HRS1_CMS0,0.0251,0.0300,0.0197,0.0222
HRS1_CMS1,0.0418,0.0500,0.0494,0.0452


Using the observed mismatch rate, the `postlink` adjustment moves the naive probabilities toward the exact-match benchmark, reducing bias from mismatched links while retaining the full sample. Results under a lower assumed mismatch rate change only modestly, suggesting that the adjustment is reasonably robust to the mismatch-rate assumption.

# 4. Brief Evaluation

We briefly evaluate the adjusted probabilities against the exact-match benchmark using several standard discrepancy and association measures.

In [15]:
# Use the adjusted table from the lower mismatch-rate sensitivity analysis
phat_adj <- adjusted_table_lower$phat

# Use the exact-match table from the aggregated data
exactprobs <- ctab_exact_prop

# Convert both tables to vectors in the same order
phat_vec <- as.vector(t(phat_adj))
exact_vec <- as.vector(t(exactprobs))

# MRAE
mrae <- mean(abs(phat_vec - exact_vec) / exact_vec)

# KLD
kld <- sum(exact_vec * log(exact_vec / phat_vec))

# GOF
n <- sum(ctab)
gof <- n * sum((exact_vec - phat_vec)^2 / exact_vec)

# Association
adj_counts <- n * phat_adj
c2 <- suppressWarnings(chisq.test(adj_counts))

# Kappa computed directly
po <- sum(diag(adj_counts)) / sum(adj_counts)
pe <- sum(rowSums(adj_counts) * colSums(adj_counts)) / sum(adj_counts)^2
kappa <- (po - pe) / (1 - pe)

# Combine results into table
evaluation_results <- data.frame(
  Metric = c("MRAE", "KLD", "GOF", "Chi-square", "Kappa"),
  Value = c(
    mrae,
    kld,
    gof,
    as.numeric(c2$statistic),
    kappa
  )
)

evaluation_results$Value <- round(evaluation_results$Value, 4)
evaluation_results

Metric,Value
<chr>,<dbl>
MRAE,0.1334
KLD,0.0017
GOF,1.0995
Chi-square,147.0581
Kappa,0.6625


Compared with the naive table, the `postlink` adjustment reduces discrepancy from the exact-match benchmark as measured by MRAE, KLD, and GOF. The lower-mismatch sensitivity analysis yields similar values, suggesting that the adjustment is reasonably stable to moderate changes in the assumed mismatch rate.